### DISCUSSION TITLE
How to prevent unnecessary use of template string?

### DISCUSSION THREAD

**Discussion by Falci** [eslint/eslint#16526] [Q&A]
We have the following code:

```jsx
render() {
  const type = `primary`;
  const name = `${this.props.name}`;
  return <Card className={`card`} type={`${type}`}>${name}</Card>
}
```

I'd like to have a rule (or rules) to prevent this unnecessary use of `` `template string` ``

This should be the correct code:
```jsx
render() {
  const type = 'primary';
  const name = this.props.name;
  return <Card className="card" type={type}>${name}</Card>
}
```

I tried to use:
```js
/*eslint quotes: ["error", "single", { "allowTemplateLiterals": false }]*/
```

It works as expected only for the `type` definition and `className` in the example.

Is there a rule for this?

[Playground](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgcXVvdGVzOiBbXCJlcnJvclwiLCBcInNpbmdsZVwiLCB7IFwiYWxsb3dUZW1wbGF0ZUxpdGVyYWxzXCI6IGZhbHNlIH1dKi9cblxuZXhwb3J0IGNvbnN0IHJlbmRlciA9ICgpID0+IHtcbiAgY29uc3QgdHlwZSA9IGBwcmltYXJ5YDtcbiAgY29uc3QgbmFtZSA9IGAke3RoaXMucHJvcHMubmFtZX1gO1xuICByZXR1cm4gPENhcmQgY2xhc3NOYW1lPXtgY2FyZGB9IHR5cGU9e2Ake3R5cGV9YH0+JHtuYW1lfTwvQ2FyZD5cbn1cblxuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6eyJqc3giOnRydWV9fSwicnVsZXMiOnsiY29uc3RydWN0b3Itc3VwZXIiOlsiZXJyb3IiXSwiZm9yLWRpcmVjdGlvbiI6WyJlcnJvciJdLCJnZXR0ZXItcmV0dXJuIjpbImVycm9yIl0sIm5vLWFzeW5jLXByb21pc2UtZXhlY3V0b3IiOlsiZXJyb3IiXSwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOlsiZXJyb3IiXSwibm8tY2xhc3MtYXNzaWduIjpbImVycm9yIl0sIm5vLWNvbXBhcmUtbmVnLXplcm8iOlsiZXJyb3IiXSwibm8tY29uZC1hc3NpZ24iOlsiZXJyb3IiXSwibm8tY29uc3QtYXNzaWduIjpbImVycm9yIl0sIm5vLWNvbnN0YW50LWNvbmRpdGlvbiI6WyJlcnJvciJdLCJuby1jb250cm9sLXJlZ2V4IjpbImVycm9yIl0sIm5vLWRlYnVnZ2VyIjpbImVycm9yIl0sIm5vLWRlbGV0ZS12YXIiOlsiZXJyb3IiXSwibm8tZHVwZS1hcmdzIjpbImVycm9yIl0sIm5vLWR1cGUtY2xhc3MtbWVtYmVycyI6WyJlcnJvciJdLCJuby1kdXBlLWVsc2UtaWYiOlsiZXJyb3IiXSwibm8tZHVwZS1rZXlzIjpbImVycm9yIl0sIm5vLWR1cGxpY2F0ZS1jYXNlIjpbImVycm9yIl0sIm5vLWVtcHR5IjpbImVycm9yIl0sIm5vLWVtcHR5LWNoYXJhY3Rlci1jbGFzcyI6WyJlcnJvciJdLCJuby1lbXB0eS1wYXR0ZXJuIjpbImVycm9yIl0sIm5vLWV4LWFzc2lnbiI6WyJlcnJvciJdLCJuby1leHRyYS1ib29sZWFuLWNhc3QiOlsiZXJyb3IiXSwibm8tZXh0cmEtc2VtaSI6WyJlcnJvciJdLCJuby1mYWxsdGhyb3VnaCI6WyJlcnJvciJdLCJuby1mdW5jLWFzc2lnbiI6WyJlcnJvciJdLCJuby1nbG9iYWwtYXNzaWduIjpbImVycm9yIl0sIm5vLWltcG9ydC1hc3NpZ24iOlsiZXJyb3IiXSwibm8taW5uZXItZGVjbGFyYXRpb25zIjpbImVycm9yIl0sIm5vLWludmFsaWQtcmVnZXhwIjpbImVycm9yIl0sIm5vLWlycmVndWxhci13aGl0ZXNwYWNlIjpbImVycm9yIl0sIm5vLWxvc3Mtb2YtcHJlY2lzaW9uIjpbImVycm9yIl0sIm5vLW1pc2xlYWRpbmctY2hhcmFjdGVyLWNsYXNzIjpbImVycm9yIl0sIm5vLW1peGVkLXNwYWNlcy1hbmQtdGFicyI6WyJlcnJvciJdLCJuby1uZXctc3ltYm9sIjpbImVycm9yIl0sIm5vLW5vbm9jdGFsLWRlY2ltYWwtZXNjYXBlIjpbImVycm9yIl0sIm5vLW9iai1jYWxscyI6WyJlcnJvciJdLCJuby1vY3RhbCI6WyJlcnJvciJdLCJuby1wcm90b3R5cGUtYnVpbHRpbnMiOlsiZXJyb3IiXSwibm8tcmVkZWNsYXJlIjpbImVycm9yIl0sIm5vLXJlZ2V4LXNwYWNlcyI6WyJlcnJvciJdLCJuby1zZWxmLWFzc2lnbiI6WyJlcnJvciJdLCJuby1zZXR0ZXItcmV0dXJuIjpbImVycm9yIl0sIm5vLXNoYWRvdy1yZXN0cmljdGVkLW5hbWVzIjpbImVycm9yIl0sIm5vLXNwYXJzZS1hcnJheXMiOlsiZXJyb3IiXSwibm8tdGhpcy1iZWZvcmUtc3VwZXIiOlsiZXJyb3IiXSwibm8tdW5kZWYiOlsiZXJyb3IiXSwibm8tdW5leHBlY3RlZC1tdWx0aWxpbmUiOlsiZXJyb3IiXSwibm8tdW5yZWFjaGFibGUiOlsiZXJyb3IiXSwibm8tdW5zYWZlLWZpbmFsbHkiOlsiZXJyb3IiXSwibm8tdW5zYWZlLW5lZ2F0aW9uIjpbImVycm9yIl0sIm5vLXVuc2FmZS1vcHRpb25hbC1jaGFpbmluZyI6WyJlcnJvciJdLCJuby11bnVzZWQtbGFiZWxzIjpbImVycm9yIl0sIm5vLXVudXNlZC12YXJzIjpbImVycm9yIl0sIm5vLXVzZWxlc3MtYmFja3JlZmVyZW5jZSI6WyJlcnJvciJdLCJuby11c2VsZXNzLWNhdGNoIjpbImVycm9yIl0sIm5vLXVzZWxlc3MtZXNjYXBlIjpbImVycm9yIl0sIm5vLXdpdGgiOlsiZXJyb3IiXSwicmVxdWlyZS15aWVsZCI6WyJlcnJvciJdLCJ1c2UtaXNuYW4iOlsiZXJyb3IiXSwidmFsaWQtdHlwZW9mIjpbImVycm9yIl19LCJlbnYiOnsiZXM2Ijp0cnVlfX19).

**Comment by mdjermanovic** [comment]
By unnecessary you mean `this.props.name` is already a string so there's no need to use `` `${this.props.name}` `` to convert it to a string?

**Reply by Falci** [reply]
Yes, unnecessary because it's not adding anything to the string.

**Reply by mdjermanovic** [reply]
`` `${foo}` `` is generally not unnecessary since it converts non-string values to strings. ESLint core rules don't have type information to know that `foo` is already a string. I'm not sure if there's a [typescript-eslint](https://github.com/typescript-eslint/typescript-eslint) rule that warns about unnecessary type conversions.

The closest functionality in core rules is [`disallowTemplateShorthand`](https://eslint.org/docs/latest/rules/no-implicit-coercion#disallowtemplateshorthand) option of the `no-implicit-coercion` rule.

[Playground demo](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgcXVvdGVzOiBbXCJlcnJvclwiLCBcInNpbmdsZVwiLCB7IFwiYWxsb3dUZW1wbGF0ZUxpdGVyYWxzXCI6IGZhbHNlIH1dKi9cbi8qZXNsaW50IG5vLWltcGxpY2l0LWNvZXJjaW9uOiBbXCJlcnJvclwiLCB7IFwiZGlzYWxsb3dUZW1wbGF0ZVNob3J0aGFuZFwiOiB0cnVlIH1dKi9cblxuZXhwb3J0IGNvbnN0IHJlbmRlciA9ICgpID0+IHtcbiAgY29uc3QgdHlwZSA9IGBwcmltYXJ5YDtcbiAgY29uc3QgbmFtZSA9IGAke3RoaXMucHJvcHMubmFtZX1gO1xuICByZXR1cm4gPENhcmQgY2xhc3NOYW1lPXtgY2FyZGB9IHR5cGU9e2Ake3R5cGV9YH0+JHtuYW1lfTwvQ2FyZD5cbn1cblxuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6eyJqc3giOnRydWV9fSwicnVsZXMiOnt9LCJlbnYiOnsiZXM2Ijp0cnVlfX19)

Note that it would autofix `` `${foo}` `` to `String(foo)`, which is probably not what you want.

If you'd like to only disallow template literals that have one `${}` and nothing else, without autofixing, you could use [`no-restricted-syntax`](https://eslint.org/docs/latest/rules/no-restricted-syntax) rule with the following selector:

```
:not(TaggedTemplateExpression) > TemplateLiteral[expressions.length=1][quasis.0.value.cooked=''][quasis.1.value.cooked='']
```

[Playground demo](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgcXVvdGVzOiBbXCJlcnJvclwiLCBcInNpbmdsZVwiLCB7IFwiYWxsb3dUZW1wbGF0ZUxpdGVyYWxzXCI6IGZhbHNlIH1dKi9cbi8qZXNsaW50IG5vLXJlc3RyaWN0ZWQtc3ludGF4OiBbXCJlcnJvclwiLCB7IFxuICAgIHNlbGVjdG9yOiBcIjpub3QoVGFnZ2VkVGVtcGxhdGVFeHByZXNzaW9uKSA+IFRlbXBsYXRlTGl0ZXJhbFtleHByZXNzaW9ucy5sZW5ndGg9MV1bcXVhc2lzLjAudmFsdWUuY29va2VkPScnXVtxdWFzaXMuMS52YWx1ZS5jb29rZWQ9JyddXCIsXG4gICAgbWVzc2FnZTogXCJVbm5lY2Vzc2FyeSB0ZW1wbGF0ZSBsaXRlcmFsLlwiXG59XSovXG5cbmV4cG9ydCBjb25zdCByZW5kZXIgPSAoKSA9PiB7XG4gIGNvbnN0IHR5cGUgPSBgcHJpbWFyeWA7XG4gIGNvbnN0IG5hbWUgPSBgJHt0aGlzLnByb3BzLm5hbWV9YDtcbiAgcmV0dXJuIDxDYXJkIGNsYXNzTmFtZT17YGNhcmRgfSB0eXBlPXtgJHt0eXBlfWB9PiR7bmFtZX08L0NhcmQ+XG59XG5cbiIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnsianN4Ijp0cnVlfX0sInJ1bGVzIjp7fSwiZW52Ijp7ImVzNiI6dHJ1ZX19fQ==)
